crates/c-api/src/func.rs

```diff
@@ -433,3 +585,116 @@ pub unsafe extern "C" fn wasmtime_func_to_raw(
 ) -> *mut c_void {
     func.to_raw(store)
 }
+
+#[cfg(feature = "async")]
+pub enum CallFutureState<'a> {
+    Undefined,
+    Called(Pin<Box<dyn Future<Output = Result<()>> + 'a>>),
+    Finished(Result<()>),
+}
+
+#[cfg(feature = "async")]
+impl Default for CallFutureState<'_> {
+    fn default() -> Self {
+        CallFutureState::Undefined
+    }
+}
+
+#[cfg(feature = "async")]
+#[repr(transparent)]
+pub struct wasmtime_call_future_t<'a> {
+    pub state: CallFutureState<'a>,
+}
+
+#[cfg(feature = "async")]
+#[no_mangle]
+pub extern "C" fn wasmtime_call_future_delete(_future: Box<wasmtime_call_future_t>) {}
+
+#[cfg(feature = "async")]
+#[no_mangle]
+pub extern "C" fn wasmtime_call_future_poll(future: &mut wasmtime_call_future_t) -> bool {
+    let mut fut = match mem::take(&mut future.state) {
+        CallFutureState::Called(fut) => fut,
+        _ => panic!("wasmtime_call_future_poll on a completed function"),
+    };
+    let w = futures::task::noop_waker_ref();
+    match fut.as_mut().poll(&mut Context::from_waker(w)) {
+        Poll::Ready(result) => {
+            future.state = CallFutureState::Finished(result);
+            true
+        }
+        Poll::Pending => {
+            future.state = CallFutureState::Called(fut);
+            false
+        }
+    }
+}
+
+#[cfg(feature = "async")]
+#[no_mangle]
+pub extern "C" fn wasmtime_call_future_get_results(
+    future: &mut wasmtime_call_future_t,
+    trap_ret: &mut *mut wasm_trap_t,
+) -> Option<Box<wasmtime_error_t>> {
+    let r = match mem::take(&mut future.state) {
+        CallFutureState::Finished(r) => r,
+        _ => panic!("wasmtime_call_future_get_results on a completed function"),
+    };
+    match r {
+        Ok(()) => None,
+        Err(err) => {
+            if err.is::<Trap>() {
+                *trap_ret = Box::into_raw(Box::new(wasm_trap_t::new(err)));
+                None
+            } else {
+                Some(Box::new(wasmtime_error_t::from(err)))
+            }
+        }
+    }
+}
+
+#[cfg(feature = "async")]
+async unsafe fn do_func_call_async(
```

Comment:
Could this, and other `async` functions, be listed as safe throughout the Rust crate? For example the main reason for the unsfaety here is the `results` and `args` raw pointers but that can be mitigated one layer up where `wasmtime_func_call_async` would perform the `slice_from_raw_parts` calls, then passing safe slices to the `async` pieces here.

I ask mainly because mixing `async` and `unsafe` is a recipe for bad things to happen, particularly around the lifetimes of arguments and such.

Reply:
Thanks yes this is better as I'm wrapping my head around async + unsafe in Rust 😄 done.